src/Microsoft.ML.LightGBM/LightGbmRankingTrainer.cs

```diff
@@ -91,16 +91,37 @@ internal LightGbmRankingTrainer(IHostEnvironment env, LightGbmArguments args)
         /// <param name="env">The private instance of <see cref="IHostEnvironment"/>.</param>
         /// <param name="labelColumn">The name of the label column.</param>
         /// <param name="featureColumn">The name of the feature column.</param>
-        /// <param name="groupIdColumn">The name for the column containing the group ID. </param>
-        /// <param name="weightColumn">The name for the column containing the initial weight.</param>
+        /// <param name="groupIdColumn">The name of the column containing the group ID. </param>
+        /// <param name="weightColumn">The name of the column containing the initial weight.</param>
+        /// <param name="numLeaves">The number of leaves to use.</param>
+        /// <param name="numBoostRound">Number of iterations.</param>
+        /// <param name="minDataPerLeaf">The minimal number of documents allowed in a leaf of the tree, out of the subsampled data.</param>
+        /// <param name="learningRate">The learning rate.</param>
         /// <param name="advancedSettings">A delegate to apply all the advanced arguments to the algorithm.</param>
-        public LightGbmRankingTrainer(IHostEnvironment env, string labelColumn, string featureColumn,
-            string groupIdColumn, string weightColumn = null, Action<LightGbmArguments> advancedSettings = null)
+        public LightGbmRankingTrainer(IHostEnvironment env,
+            string labelColumn,
+            string featureColumn,
+            string groupIdColumn,
+            string weightColumn = null,
+            int? numLeaves = null,
+            int? minDataPerLeaf = null,
+            double? learningRate = null,
+            int numBoostRound = LightGbmArguments.Defaults.NumBoostRound,
+            Action<LightGbmArguments> advancedSettings = null)
             : base(env, LoadNameValue, TrainerUtils.MakeR4ScalarLabel(labelColumn), featureColumn, weightColumn, groupIdColumn, advancedSettings)
         {
             Host.CheckNonEmpty(labelColumn, nameof(labelColumn));
             Host.CheckNonEmpty(featureColumn, nameof(featureColumn));
             Host.CheckNonEmpty(groupIdColumn, nameof(groupIdColumn));
+
+            if (advancedSettings != null)
+                CheckArgsAndAdvancedSettingMismatch(numLeaves, minDataPerLeaf, learningRate, numBoostRound, new LightGbmArguments(), Args);
+
```

Comment:
I think there has been a misconception. 

It is OK to have the parameters (like `numLeaves` and `learningRate`) available in 2 places (inside `Args` and as a ctor argument.
The correct pattern is
```csharp
var args = new Arguments();
args.NumLeaves = numLeaves;
args.LearningRate  = learningRate;
//...
advancedSettings?.Invoke(args);
```
That is, we set the args' fields to the user-provided ctor arguments, and then let the user optionally change them again in the delegate: this is not a bug.

One exception is for column names: we are always ignoring the column names if they are set in the arguments. So I suggested to check (and throw) if the user sets something like `FeatureColumn` in the args to the value different to `featureColumn` ctor argument (as we will not respect that change).

Apparently, the latter is not yet done, and it's probably fine for now. But let's not check the other args.

Reply:
what I have been doing, is the exact opposite, and i do prefer that tbh because of two reasons:

1- We do have parms in two places, and i think  whenever we can deprecate those params from the arguments, we should do so. I don;t think we should promote the buried arguments in favor of the ones that  surfaced in the signature. 

2- It keeps the behavior consistent with the column names.

cc @tomfinley for his opinion. 

---
In reply to: [227869641](https://github.com/dotnet/machinelearning/pull/1352#discussion_r227869641) [](ancestors = 227869641)

Reply:
I am not a fan of enforcing a mismatch, since I could easily imagine situations where someone sets some reasonable arguments, but *might* thanks to some complicated logic decide to think better of it in the more complex setter.

However there is an addition point I wanted to address, since my first reaction is to disagree with it somewhat.

> We do have parms in two places, and i think  whenever we can deprecate those params from the arguments, we should do so.

I am strongly not in favor of this. I believe it is essential that the object with the advanced setting also continue to have the basic settings.

The reason is, a delegate for setting the object *might* be something used alongside the constructor, but in more programmatic, non-direct instantiation settings, especially in situations where several decisions have to be taken, it might be a real honest-to-god separate method living in a class somewhere. If the options argument has omitted the most common options, as suggested here, then we would force this scenario to spread its setting objects across two places.

A more minor consideration is that it seems more in-line with the [idiom from ASP.NET](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration/options?view=aspnetcore-2.1), which is where @alexdegroot originally suggested we lift the idiom [here](https://github.com/dotnet/machinelearning/issues/581#issuecomment-408932526).

My inclination therefore is to agree with @Zruty0 on this one, sorry @sfilipi. :)

---
In reply to: [227886248](https://github.com/dotnet/machinelearning/pull/1352#discussion_r227886248) [](ancestors = 227886248,227869641)

Reply:
thanks for the elaborate and convincing argument @tfinley@gmail.com. changing it everywhere. 

---
In reply to: [228319239](https://github.com/dotnet/machinelearning/pull/1352#discussion_r228319239) [](ancestors = 228319239,227886248,227869641)